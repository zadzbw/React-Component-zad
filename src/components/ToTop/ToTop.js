/**
 * Created by zad on 16/12/13.
 */
import React, {PropTypes, Component} from 'react';
import Animate from 'rc-animate';
import classNames from 'classnames';
import _omit from 'lodash/omit';
import getScroll from '../../utils/getScroll';
import Motion from '../../utils/Motion';
import Icon from '../Icon';

import './ToTop.less';

function quarticInOut(t, start, end, duration) {
  const diff = end - start;
  // 归一化
  t = t / duration;
  return start + diff * Motion.quarticInOut(t);
}

export default class ToTop extends Component {
  constructor(props) {
    super(props);
    this.state = {
      visible: false
    };
    this._scrollToTop = this._scrollToTop.bind(this);
    this._setTopValue = this._setTopValue.bind(this);
    this._listenScroll = this._listenScroll.bind(this);
  }

  static propTypes = {
    triggerHeight: PropTypes.number,
    duration: PropTypes.number,
    target: PropTypes.func,
    onClick: PropTypes.func
  };

  static defaultProps = {
    triggerHeight: 300,
    duration: 400,
    target: () => window,
    onClick: () => undefined
  };

  _scrollToTop(e) {
    if (window.requestAnimationFrame) {
      let start = null;
      const {duration, target} = this.props;
      // 获取 target 的滚动高度
      const scrollTop = getScroll(target()).top;
      const step = (now) => {
        if (!start) {
          start = now;
        }
        const diff = now - start;
        this._setTopValue(quarticInOut(diff, scrollTop, 0, duration));
        if (diff < duration) {
          window.requestAnimationFrame(step);
        }
      };
      window.requestAnimationFrame(step);
    } else {
      window.scroll(0, 0);
    }
    this.props.onClick(e);
  }

  _setTopValue(value) {
    const target = (this.props.target)();
    if (target === window) {
      document.body.scrollTop = value; // Safari and Chrome
      document.documentElement.scrollTop = value; // Firefox Opera and IE
    } else if (target instanceof HTMLElement) {
      target.scrollTop = value;
    }
  }

  _listenScroll() {
    const {triggerHeight, target} = this.props;
    const scrollTop = getScroll(target()).top;
    this.setState({
      visible: scrollTop > triggerHeight
    });
  }

  componentDidMount() {
    const {target} = this.props;
    target().addEventListener('scroll', this._listenScroll, false);
  }

  componentWillUnmount() {
    const {target} = this.props;
    target().removeEventListener('scroll', this._listenScroll, false);
  }

  render() {
    const {className, children, ...rest} = this.props;
    const props = _omit(rest, ['duration', 'target', 'triggerHeight', 'onClick']);
    const defaultContent = (
      <div className="zad-to-top-content">
        <Icon className="zad-to-top-icon" name={'arrow-up'}/>
      </div>
    );
    const toTopClass = classNames('zad-to-top', className);

    const toTopBtn = !this.state.visible ? null : (
      <div {...props} className={toTopClass} onClick={this._scrollToTop}>
        {children || defaultContent}
      </div>
    );

    return (
      <Animate transitionName="zoom">
        {toTopBtn}
      </Animate>
    );
  }
}
