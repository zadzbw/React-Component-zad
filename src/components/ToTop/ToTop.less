@import "../../style/variables";

.zad-to-top {
  position: fixed;
  right: 100px;
  bottom: 50px;
  height: 40px;
  width: 40px;
  cursor: pointer;
  z-index: @z-index-to-top;

  &-content {
    height: 40px;
    width: 40px;
    border-radius: 50%;
    color: @body-background;
    background-color: hsla(0, 0%, 25%, 0.4);
    text-align: center;
    transition: all @animation-duration-slow @ease-in-out;

    &:hover {
      background-color: hsla(0, 0%, 25%, 0.6);
      transition: all @animation-duration-slow @ease-in-out;
    }
  }

  &-icon {
    font-size: 20px;
    margin-top: 10px;
  }
}
