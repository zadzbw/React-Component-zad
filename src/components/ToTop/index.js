/**
 * Created by zad on 16/12/13.
 */
import ToTop from './ToTop';

export default ToTop;
