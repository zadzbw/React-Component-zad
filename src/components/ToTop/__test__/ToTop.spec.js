/**
 * Created by zad on 16/12/13.
 */
import React from 'react';
import {mount} from 'enzyme';
import {mountToJson} from 'enzyme-to-json';
import ToTop from '../../ToTop';

describe('ToTop test', () => {
  it('init', () => {
    const wrapper = mount(
      <ToTop/>
    );

    expect(wrapper.text()).toEqual('');
    expect(wrapper.find('div')).toHaveLength(0);
    expect(mountToJson(wrapper)).toMatchSnapshot();
  });
});
