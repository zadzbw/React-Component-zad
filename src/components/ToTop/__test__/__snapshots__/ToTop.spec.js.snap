exports[`ToTop test init 1`] = `
<ToTop
  duration={400}
  onClick={[Function]}
  target={[Function]}
  triggerHeight={300}>
  <Animate
    animation={Object {}}
    component="span"
    onAppear={[Function]}
    onEnd={[Function]}
    onEnter={[Function]}
    onLeave={[Function]}
    transitionAppear={false}
    transitionEnter={true}
    transitionLeave={true}
    transitionName="zoom">
    <span />
  </Animate>
</ToTop>
`;
